docs(MADR): create MADR for MeshTrafficPermission (#4666)
kumahq/kuma: docs/madr/decisions/006-mesh-traffic-permission.md

## Decision Outcome

Chosen option: create a MeshTrafficPermission policy

### Overview

MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
There are might be several reasons why it could be useful to block unwanted traffic:
* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)

### Specification

Top-level targetRef can have all available kinds:

```yaml
targetRef:
  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
  name: ...
```

MeshTrafficPermission is an inbound policy that's why it has a "from" array:

```yaml
from:
  - targetRef:
      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
      name: ...
```

Configuration for MeshTrafficPermission has a single field "action" with 2 values "ALLOW" or "DENY".
So the overall specification looks like this:

```yaml
type: MeshTrafficPermission
mesh: default
spec:
  targetRef:
    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
    name: ...
  from:

>     - targetRef:

Reviewer: how do we handle cross-mesh permissions?

```
fromMesh: other-mesh
from:
- kind: MeshService
  name: service-name
```
?

Author: Are there cases when you want to combine regular permissions with cross-mesh permissions in one policy? Maybe `MeshService` targetRef should have a `mesh` field, which is empty by default. But if you set it, then it's the service from another mesh:
```yaml
from:
  - kind: MeshService
    mesh: other-mesh
    name: service-name
  - kind: MeshService
    name: backend
```

Reviewer: I don't think anyone would _need_ that option. You can always just create two policies I think. The `targetRef` option is there of course, but maybe it's cleaner to add it to the policy spec since I think it only works for `MeshTrafficPermission`.

Author: What happens when you concatenate `from` arrays from different levels? Do you have separate `from` arrays for cross mesh and regular permissions?

Reviewer: I think we can just treat it internally as if there _were_ a `mesh` field in the `targetRef` entries.

Then again, this is really only a thing for `targetRef` kinds that aren't real objects so maybe it should just be a field for those kinds and we just use validation to enforce these rules.

Author: We can create a `mesh` field in the "from" item:
```yaml
type: MeshTrafficPermission
mesh: expose
name: cross-mesh-permission
spec:
  targetRef:
    kind: MeshGatewayRoute
    name: cross-mesh-route
  from:
    - mesh: consume
      targetRef:
        kind: MeshService
        name: backend
      conf:
        action: ALLOW
    - mesh: consume
      targetRef:
        kind: MeshServiceSubset
        name: backend
        tags:
          version: v1
      conf:
        action: DENY
```
so only MeshTrafficPermission will have `mesh`. Other inbound policies with "from" arrays (FaultInjection or RateLimit) won't have it

Reviewer: Ah of course! Nice

Reviewer 2: Who is doing a great job on the team? What have they done?

Reviewer 2: What's the reason to have this at the same level as `targetRef` rather than being a field in the `targetRef`? I'm not sure I understand why we gave up on this

Author: Added "mesh" into targetRef

Reviewer: @lahabana you mean gave up on it being a field in `targetRef`? I was just concerned that it may not be applicable/possible for every policy but putting it in `targetRef` would apply to every policy. But maybe it is possible for every policy or eventually will be?

        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
        name: ...
      conf:

>         action: ALLOW|DENY

Reviewer 2: Should there be an option `shadow: true` to run this RBAC rule in shadow (get stats but no enforcement)?

Author: Added "shadow_action" field

Reviewer 3: would it be possible to have
```
action: ALLOW|DENY|SHADOW_ALLOW|SHADOW_DENY
```
?

Reviewer 3: Is there any reason to set `action: ALLOW` + `shadow: ALLOW` or `action: DENY` + `shadow: DENY`?

Reviewer 2: While it doesn't make sense. I think infering that `SHADOW_ALLOW` means action `DENY` is a little counter intuitive no?

Reviewer: Would we always require both to be set?

Otherwise

```
default:
  shadowAction: ALLOW
```

would also be valid and mean `action: DENY` so it wouldn't be any less counter intuitive.

Author: hmm, but I think it doesn't mean `action: DENY` it means there is no real rule for this set of clients. If you don't have a rule, then yes, the request will be denied, but don't generate rules for that

Reviewer 4: > Is there any reason to set action: ALLOW + shadow: ALLOW or action: DENY + shadow: DENY?

`action: ALLOW + shadow: ALLOW` => want traffic to flow and you don't want shadow engine stats

`action: DENY + shadow: DENY` => want to stop traffic and you want metrics

It all depends on how this translates down to Envoy, if it's a 1:1 mapping it's as I described above.

Author: if I understand envoy docs correctly, by default, stats are the same regardless of the `shadow`. But you can add a prefix to shadow stats using [shadow-rules-stat-prefix](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/rbac/v3/rbac.proto#envoy-v3-api-field-extensions-filters-network-rbac-v3-rbac-shadow-rules-stat-prefix).

Reviewer: > action: ALLOW + shadow: ALLOW => want traffic to flow and you don't want shadow engine stats
> action: DENY + shadow: DENY => want to stop traffic and you want metrics

Do you know why would someone configure this? You get stats from both actions and shadow actions.

```

We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 

### Envoy configuration

Envoy has [RBAC Network Filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/rbac_filter) 
that allows configuring permissions for the list of principals.

**Principal** is an entity that can be authenticated.<br>
**Permission** is access granted to the principal.  

Each Kuma service has a generated certificate with URIs that identify service name and tags.
For example "backend" service in "us-east" zone of version v1 can have the following list of URIs:
* spiffe://mesh-1/backend
* kuma://kuma.io/zone/us-east
* kuma://version/v1

If we want to grant any permissions to these instances we can create the following Network RBAC:

```yaml
action: ALLOW
policies:

>   "backend-us-east-v1-allow-all":

Reviewer 2: The policyName is always the MeshTrafficPermission resource name it comes from?

If yes you should document that in the MADR

Author: It's the name of the most specific MeshTrafficPermission for the DPP. Added this to MADR

Reviewer 2: I thought we settled on always having `MeshTrafficPermission` no?

Reviewer 2: Still seem to be here no?

Author: Missed this place, fixed now

```

In order request to be allowed, at least 1 policy should be matched. 

#### Converting MeshTrafficPermission to Envoy NetworkRBAC

1. Concatenate "from" arrays from all matched MeshTrafficPermissions 

Input
```yaml
type: MeshTrafficPermission
name: global
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: Mesh
      conf:
        action: DENY
    - targetRef:
        kind: MeshSubset
        tags:
          kuma.io/zone: us-east
      conf:
        action: ALLOW
---
type: MeshTrafficPermission
name: backend
spec:
  targetRef:
    kind: MeshService
    name: backend
  from:
    - targetRef:
        kind: MeshSubset
        tags:
          env: dev
      conf:
        action: DENY
```

Concatenated "from" array
```yaml
from:
  - targetRef:
      kind: Mesh
    conf:
      action: DENY
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
    conf:
      action: ALLOW
  - targetRef:
      kind: MeshSubset
      tags:
        env: dev
    conf:
      action: DENY
```

2. Create a full rule-based view (with negations) 

Regular [rule-based view](005-policy-matching.md#rule-based-view) doesn't work well for NetworkRBAC generation,

> because rule-based view relies on the order, but NetworkRBAC doesn't. 

Reviewer 3: But it can rely on order, no?
"The policies are evaluated in lexicographic order of the policy name"
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-msg-config-rbac-v3-rbac
we pick a name for the policies in NetworkRBAC filter. We can sort them then however we want.

Reviewer 3: I see. Ok.

Reviewer 4: I remember order being important when rules overlap. E.g.

allow: /* 
deny: /some-path/*

The `deny` in this case is useless. Have you taken into account the overlapping of rules?

In this example, if a request matches "rule 2" this means it doesn't match "rule 1":
```yaml
rules:
  - targetRef: # rule 1
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: dev
    conf:
      action: DENY
  - targetRef: # rule 2
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
    conf:
      action: ALLOW
```

So "rule 2" implicitly matches requests from zone "us-east" that are **not** from env "dev".
That's why it's important to have a full rule-based view with negations ("!" means "not"):

```yaml
rules:
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: dev
    conf:
      action: DENY
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: !dev
    conf:
      action: ALLOW
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: !us-east
        env: dev
    conf:
      action: DENY 
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: !us-east
        env: !dev
    conf:
      action: DENY
```

> See [how to generate a full rule-based view with negations](#how-to-generate-a-full-rule-based-view-with-negations)

Reviewer 2: Do we agree that this "rule-based" view with negations is only an internal and conceptual thing used here to explain how we'll do the mapping, It will never be exposed to the user?

Author: Yes, we don't need to expose that to users

3. For each "rule" that has "ALLOW" action generate a separate NetworkRBAC policy

```yaml

> action: ALLOW

Reviewer 2: So the action will always be ALLOW? regardless on the policies?

Author: Yes, it's always ALLOW

policies:
  "mtp-1":
    permissions:
      - any: true
    principals:
      - and_ids:
          ids:
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
```

4. If MeshTrafficPermission is targeting proxy then NetworkRBAC is placed as a NetworkFilter on the inbound listener.

> If MeshTrafficPermission is targeting MeshHTTPRoute then NetworkRBAC is placed as a NetworkFilter on the route.

Reviewer 2: How will your "rule-based view with negations" work in this case? Do you just have 2 different ones one for route and one for listener?

Author: That's correct, there is per-listener and per-route RBAC. Added this to MADR

### MeshTrafficPermission and Kubernetes

MeshTrafficPermission is going to be the first policy with a new policy matching. 
There are a few things worth mentioning regarding this fact. 

#### Mesh as label

Instead of creating a CRD with a "mesh" field, we can use a label to store mesh:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  labels:
    kuma.io/mesh: mesh-1
```

> This allows filtering policies by mesh using kumactl.

Reviewer 2: Mention that if the label is unset it will be added by the webhook to be "default"

Reviewer 2: ```suggestion
This allows filtering policies by mesh using kubectl.
```

#### Scope is namespaced

Today all policies are cluster-scoped, but new policies should be namespace scoped.

> Policy applies only to the DPPs of the same namespace.

Reviewer 2: I don't think we need to do this now (let's keep the scope small). So for the moment policies can only be created in the namespace the CP runs in

3. For each "rule" that has "ALLOW" action generate a separate NetworkRBAC policy

```yaml
action: ALLOW
policies:
  "mtp-1":
    permissions:
      - any: true
    principals:
      - and_ids:
          ids:
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
```

4. If MeshTrafficPermission is targeting proxy then NetworkRBAC is placed as a NetworkFilter on the inbound listener.
If MeshTrafficPermission is targeting MeshHTTPRoute then NetworkRBAC is placed as a NetworkFilter on the route.

### MeshTrafficPermission and Kubernetes

MeshTrafficPermission is going to be the first policy with a new policy matching. 
There are a few things worth mentioning regarding this fact. 

#### Mesh as label

Instead of creating a CRD with a "mesh" field, we can use a label to store mesh:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  labels:
    kuma.io/mesh: mesh-1
```

> This allows filtering policies by mesh using kumactl. 

Reviewer 2: ```suggestion
This allows filtering policies by mesh using kubectl. 
```

because rule-based view relies on the order, but NetworkRBAC doesn't. 

In this example, if a request matches "rule 2" this means it doesn't match "rule 1":
```yaml
rules:
  - targetRef: # rule 1
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: dev
    default:
      action: DENY
  - targetRef: # rule 2
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
    default:
      action: ALLOW
```

So "rule 2" implicitly matches requests from zone "us-east" that are **not** from env "dev".
That's why it's important to have a full rule-based view with negations ("!" means "not"):

```yaml
rules:
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: dev
    default:
      action: DENY
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: !dev
    default:
      action: ALLOW
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: !us-east
        env: dev
    default:
      action: DENY 
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: !us-east
        env: !dev
    default:
      action: DENY

> ```

Reviewer 2: Shouldn't we trim all DENY rules in this step as well (for clarity)?

Author: Added an extra step for that

See [how to generate a full rule-based view with negations](#how-to-generate-a-full-rule-based-view-with-negations)

3. For each "rule" that has "action: ALLOW" generate a single "principal" in v3.RBAC:

```yaml
action: ALLOW
policies:
  "backend": # name of the most specific MeshTrafficPermission policy for the DPP
    permissions:
      - any: true
    principals: # each principal is matched with OR semantics
      - and_ids:
          ids:
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"

> ```

Reviewer 2: Didn't we talked about a policy with "or_ids" at the top level?

Author: I missed this fact before that 1 policy has a **list** of principals, and principals are matched with OR semantic. So I added `For each "rule" that has "action: ALLOW" generate a single "principal" in v3.RBAC`.

Reviewer 2: Could this example be expand to something less trivial then? We're trying to capture in a MADR the decisions and the reasons we're taking these decisions.

Author: slightly improved the example to have at least 2 "allow" rule at this step

Similarly generate a v3.RBAC using "shadow_action: ALLOW". 
Generated RBACs are set in the network filter as "rules" and "shadow_rules" respectively.

4. If MeshTrafficPermission is targeting proxy then NetworkRBAC is placed as a NetworkFilter on the inbound listener.
If MeshTrafficPermission is targeting MeshHTTPRoute then NetworkRBAC is placed as a NetworkFilter on the route. 
Even in the case of MeshHTTPRoute, global per-proxy NetworkRBAC is still will be set to provide secure access for other routes. 

### MeshTrafficPermission and Kubernetes

MeshTrafficPermission is going to be the first policy with a new policy matching. 
There are a few things worth mentioning regarding this fact. 

#### Mesh as label

Instead of creating a CRD with a "mesh" field, we can use a label to store mesh:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  labels:
    kuma.io/mesh: mesh-1
```

This allows filtering policies by mesh using kubectl. 
If the label is unset it will be added automatically by the webhook to be "kuma.io/mesh: default".

#### Scope is namespaced

Today all policies are cluster-scoped, but new policies should be namespace scoped.

> MeshTrafficPermission is limited to have only "kuma-system" namespace, 

Reviewer 2: ```suggestion
MeshTrafficPermission can only be created in the namespace where the control-plane runs (`kuma-system` by default), 
```

and affects DPPs across all namespaces.

> In the future we'll add support for other namespaces, so the policy could affect DPPs only in the specified namespace. 

Reviewer 2: ```suggestion
In the future we'll add support for other namespaces, so the policy could affect DPPs only in the specified namespace but this is out of scope for this MADR. 
```

### Cross-mesh

MeshTrafficPermission is used to specify a list of clients from other meshes that have an access to MeshGatewayRoute.
In that case we use targetRef with "mesh" field:

```yaml
type: MeshTrafficPermission
mesh: expose
name: cross-mesh-permission
spec:
  targetRef:
    kind: MeshGatewayRoute
    name: cross-mesh-route
  from:
    - targetRef:
        kind: MeshService
        name: backend
        mesh: consume
      default:
        action: ALLOW
    - targetRef:
        kind: MeshServiceSubset
        name: backend
        mesh: consume
        tags:
          version: v1
      default:
        action: DENY

> ```

Reviewer 2: Feels like the mesh field should also be added in [005](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md). It should also be clear in the addition to 005 that having `mesh` unset means same mesh as the current policy and that setting a different mesh is dependent on the policy type and the main `targetRef`. 
These are details that are important to articulate as it's part of our decision!

Author: I added it yesterday, it was in the PR that changed `conf` to `default` https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md?plain=1#L66

    {key: "env", value: "prod"},
}
comb8 = []Tag{ // no clients can have env=dev and env=prod at the same time
    {key: "zone", value: "!us-east"},
    {key: "env", value: "dev"},
    {key: "env", value: "prod"},
}
```

![img.png](assets/rules-example.png)

3. Determine configuration for each combination (only number 5-green is DENY):

```yaml
rules:
  - targetRef:
      kind: MeshSubset
      tags:
         zone: us-east
         env: dev
    default:
      action: ALLOW
  - targetRef:
       kind: MeshSubset
       tags:
          zone: us-east
          env: prod
    default:
       action: ALLOW
  - targetRef:
       kind: MeshSubset
       tags:
          zone: !us-east
          env: dev
    default:
       action: ALLOW
  - targetRef:
       kind: MeshSubset
       tags:
          zone: !us-east
          env: prod
    default:
       action: ALLOW
  - targetRef:
       kind: MeshSubset
       tags:
          zone: us-east
          env: !prod
          env: !dev
    default:
       action: DENY
  - targetRef:
       kind: MeshSubset
       tags:
          zone: !us-east
          env: !prod
          env: !dev
    default:
       action: ALLOW

> ```

Reviewer 2: I think it'd be useful to add a step that drop all the `DENY` entries as they are not needed anymore.

Author: Done

[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
Current MADR aims to define a MeshTrafficPermission.

## Considered Options

* Create a MeshTrafficPermission policy

## Decision Outcome

Chosen option: create a MeshTrafficPermission policy

### Overview

MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
There are might be several reasons why it could be useful to block unwanted traffic:
* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)

### Specification

Top-level targetRef can have all available kinds:

```yaml
targetRef:
  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
  name: ...
```

MeshTrafficPermission is an inbound policy that's why it has a "from" array:

```yaml
from:
  - targetRef:
      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
      name: ...
```

> Configuration for MeshTrafficPermission has 2 fields "action" and "shadow_action" with values "ALLOW" or "DENY".

Reviewer: Of course it can't be "shadow_action" but "shadowAction".

Author: Oh right, that's true

mesh: default
spec:
  targetRef:
    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
    name: ...
  from:
    - targetRef:
        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
        name: ...
      default:
        action: ALLOW|DENY
        shadow_action: ALLOW|DENY
```

We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 

### Shadow actions

Shadow actions are not enforced but emit stats and can be used for rule testing.
For example, before creating "action: DENY" rule, user creates "shadow_action: DENY" rule and checks the stats.
If there are no unexpected denied requests from the clients then it's safe to change "shadow_action" to "action".

Alternatively we could introduce `shadow: true|false` instead of `shadow_action: ALLOW|DENY`. 
But this approach has its flaws:
* policy matching makes it less obvious with merging. It's easy to create a rule like:
    ```yaml
    targetRef:
       kind: Mesh
    default:
      action: DENY
      shadow: true
   ```

>    and unintentionally shadow all less specific rules.

Reviewer: How is it easier to do this vs 

```yaml
targetRef:
  kind: Mesh
default:
  shadowAction: DENY
```
?

Author: because you won't affect `action: DENY` of the less specific rules. But with `shadow: true` you can affect `action` field of other rules (because fields are correlated)

Reviewer: That's only because we previously defined a rigid rule about merging top level fields that applies to _all policies_. It's unintentionally confusing because we interpret the semantics as if it were:

```
default:
  rule:
    action: ALLOW|DENY
    shadow: true
```

Anyway, action/shadowAction can just as easily lead to unintentionally overriding less specific rules in exactly the same way, can't it?

=> it should be `action: ALLOW|DENY|SHADOW_DENY|SHADOW_ALLOW`.

Reviewer: > Anyway, action/shadowAction can just as easily lead to unintentionally overriding less specific rules in exactly the same way, can't it?

That is to say, they're not orthogonal to one another either:

```
type: MeshTrafficPermission
name: global
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
         kind: MeshService
         name: backend
       default:
         shadowAction: ALLOW
---
type: MeshTrafficPermission
name: mtp-web
spec:
  targetRef:
    kind: MeshService
    name: web
  from:         
    - targetRef:
         kind: Mesh
       default:
         action: ALLOW
```

and now I have

```
action: ALLOW
shadowAction: ALLOW
```

when the user probably wanted:

```
action: DENY
shadowAction: ALLOW
```

They forgot to set `action: DENY` in the more specific policy just like they forgot to set `shadow: false` in the more specific policy in the `shadow: bool` example.

Reviewer 2: Except it's possible to differentiate a field unset from a boolean unset (as it's false). So when you don't set a field lower down it uses whatever is inherited from the parent. This is impossible with booleans right?

Reviewer: I'm not sure what you mean but `bool <=> {ALLOW, DENY}` and `*bool <=> *{ALLOW, DENY}`. There's no difference, is there?

Author: yeah, we could use `google.protobuf.BoolValue` to distinguish `bool` and `nil`. But I guess it doesn't matter since we don't want `shadow: true|false` anyway

* Envoy configuration has 2 independent places for "rules" and "shadow_rules", 

> so having independent fields "action" and "shadow_action" makes it easier to translate to Envoy configs.

Reviewer: This doesn't make sense to me. is it hard either way? Why is one way easier?

Author: If you have `action: ALLOW|DENY` and `shadow: true|false`, then these fields are sort of correlated (changing of the `shadow` flag changes the meaning of `action`). You can't independently generate "rules" and "shadow_rules", at least it's not simple.

But if you have `action: ALLOW|DENY` and `shadow_action: ALLOW|DENY`, then these fields are not correlated. And you can generate "rules" and "shadow_rules" independently. And that's exactly how Envoy has it in the configuration.

Does it make sense? I can rephrase the sentence in the document

Reviewer: You can always convert the `{shadow: bool, action: ALLOW|DENY}` model into the `{shadowAction: ALLOW|DENY, action: ALLOW|DENY}` model, so it isn't any simpler or harder, right?

Do you mean that you can't express something like `{action: ALLOW, shadowAction: DENY}` in one entry with the `shadow: bool` model?

Reviewer: In fact I agree with @jakubdyszkiewicz that `ALLOW|DENY|SHADOW_ALLOW|SHADOW_DENY` would be the simplest. `{action, shadowAction}` is too unconstrained in my opinion and allows easily misconfiguring something like `{action: DENY, shadowAction: DENY}`

Author: That's true, you can convert, but the boolean flag makes it difficult when merging. For example:
```yaml
type: MeshTrafficPermission
name: global
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
         kind: MeshService
         name: backend
       default:
         action: DENY
---
type: MeshTrafficPermission
name: mtp-web
spec:
  targetRef:
    kind: MeshService
    name: web
  from:         
    - targetRef:
         kind: Mesh
       default:
         action: DENY
         shadow: true
```
in this example `shadow: true` from `mtp-web` changes the behaviour of `action: DENY` in `global`. And in fact you have no actual DENY rules in this case, only shadow rules

Reviewer: But how does that make it "difficult", it's just easy to misconfigure, right? That's the same as the following "unintentionally shadow rules" point.

Reviewer: that's just `action: SHADOW_DENY`, isn't it?

Reviewer: Resolving this in favor of my third comment where we end up discussing the same topic.

      default:
        action: ALLOW
---
type: MeshTrafficPermission
name: backend
spec:
  targetRef:
    kind: MeshService
    name: backend
  from:
    - targetRef:
        kind: MeshSubset
        tags:
          env: dev
      default:
        action: DENY
```

Concatenated "from" array
```yaml
from:
  - targetRef:
      kind: Mesh
    default:
      action: DENY
  - targetRef:
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
    default:
      action: ALLOW
  - targetRef:
      kind: MeshSubset
      tags:
        env: dev
    default:
      action: DENY
```

2. Create a full rule-based view (with negations) 

Regular [rule-based view](005-policy-matching.md#rule-based-view) doesn't work well for NetworkRBAC generation,
because rule-based view relies on the order, but NetworkRBAC doesn't. 

In this example, if a request matches "rule 2" this means it doesn't match "rule 1":
```yaml
rules:
  - targetRef: # rule 1
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
        env: dev
    default:
      action: DENY
  - targetRef: # rule 2
      kind: MeshSubset
      tags:
        kuma.io/zone: us-east
    default:
      action: ALLOW
```

> So "rule 2" implicitly matches requests from zone "us-east" that are **not** from env "dev".

Reviewer: Why, isn't this something we decide? Isn't this a problem in general with `MeshSubset` and "specificity" of tags?

What's the problem with making the order matter here?

If `env: dev` is first, then `env: dev` requests are allowed and `!env: dev` requests are allowed.
If `env: dev` is second, then the rule is essentially ignored?

Author: Discussed offline, can we resolve this?

```

This allows filtering policies by mesh using kubectl. 
If the label is unset it will be added automatically by the webhook to be "kuma.io/mesh: default".

#### Scope is namespaced

Today all policies are cluster-scoped, but new policies should be namespace scoped.
MeshTrafficPermission can only be created in the namespace where the control-plane runs (`kuma-system` by default), 
and affects DPPs across all namespaces.
In the future we'll add support for other namespaces, so the policy could affect DPPs only in the specified namespace but this is out of scope for this MADR. 

### Cross-mesh

MeshTrafficPermission is used to specify a list of clients from other meshes that have an access to MeshGatewayRoute.
In that case we use targetRef with "mesh" field:

```yaml
type: MeshTrafficPermission
mesh: expose
name: cross-mesh-permission
spec:
  targetRef:
    kind: MeshGatewayRoute
    name: cross-mesh-route
  from:
    - targetRef:
        kind: MeshService
        name: backend
        mesh: consume
      default:
        action: ALLOW
    - targetRef:
        kind: MeshServiceSubset
        name: backend
        mesh: consume
        tags:
          version: v1
      default:

>         action: DENY

Reviewer 2: So how will this work with all your negations btw? I guess it work there's just a "special" match for `targetRef: kind Mesh name: consume` where the principal matching is prefixed based right?

Author: When we generate a rule-based view with negations, we can add an artificial `kuma.io/mesh` tag. So we should look at 2 cases: `kuma.io/mesh: consume` and `kuma.io/mesh: !consume`

[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
Current MADR aims to define a MeshTrafficPermission.

## Considered Options

* Create a MeshTrafficPermission policy

## Decision Outcome

Chosen option: create a MeshTrafficPermission policy

### Overview

MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
There might be several reasons why it could be useful to block unwanted traffic:
* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)

### Specification

Top-level targetRef can have all available kinds:

```yaml
targetRef:
  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
  name: ...
```

MeshTrafficPermission is an inbound policy that's why it has a "from" array:

```yaml
from:
  - targetRef:
      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
      name: ...
```

> Configuration for MeshTrafficPermission has a single field "action" with values "ALLOW", "DENY", 

Reviewer 4: There is now an action called log - https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto.html?highlight=action%20log#envoy-v3-api-enum-config-rbac-v3-rbac-action are we considering it in any way?

> "ALLOW_WITH_SHADOW_DENY" and "DENY_WITH_SHADOW_ALLOW".

Reviewer 4: what does `DENY_WITH_SHADOW_ALLOW` do?

Author: it generates stats as if the request is allowed, but practically it's denied

Reviewer 4: So why would that be useful? The request is denied and I get stats as if it went through, I don't see the value in that. The other way around I get, you allow the request to go through but generate stats as if the action was denied.

Reviewer: Isn't it just the dual of that? I mean `ALLOW_WITH_SHADOW_DENY` is useful when you're moving from `ALLOW` -> `DENY`, I think, so `DENY_WITH_SHADOW_ALLOW` is for moving from `DENY` to `ALLOW` . Or put another way, why have shadow allow at all if `DENY_WITH_SHADOW_ALLOW` isn't useful?

Reviewer 4: > DENY_WITH_SHADOW_ALLOW is for moving from DENY to ALLOW

Yeah, probably right. I was under impression that it might be less important (/ not needed at all) since we're going from `DENY` to `ALLOW`. People would be very careful restricting traffic but I thought they might not care that much when loosening up the restrictions but I guess it depends on the change.

So the overall specification looks like this:

```yaml
type: MeshTrafficPermission
mesh: default
spec:
  targetRef:
    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
    name: ...
  from:
    - targetRef:
        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
        name: ...
      default:
        action: ALLOW|DENY|ALLOW_WITH_SHADOW_DENY|DENY_WITH_SHADOW_ALLOW
```

We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 

### Shadow actions

Shadow actions are not enforced but emit stats and can be used for rule testing.
For example, before creating "action: DENY" rule, user creates "action: ALLOW_WITH_SHADOW_DENY" rule and checks the stats.

> If there are no unexpected denied requests from the clients then it's safe to change "action" to "DENY".

Reviewer 4: Comment: If there are, what does the user do then? For this feature to be useful it needs to be coupled with `access_log_hint` (+ maybe some other mechanisms) so that the user can identify the caller.

Author: Yeah, right now, it's just an indication some clients request the service. Users should set access logs or check application logs to figure out who are these clients

#### Other considered options

Option 1: `action: ALLOW|DENY, shadow: true|false`

Cons: policy matching makes it less obvious with merging. It's easy to create a rule like:
```yaml
targetRef:
   kind: Mesh
default:
  action: DENY
  shadow: true
```
and unintentionally shadow all less specific rules.

Option 2: `action: ALLOW|DENY, shadowAction: ALLOW|DENY`

Cons:

> * there are unused states like `action: ALLOW, shadowAction: ALLOW` and `action: DENY, shadowAction: DENY`

Reviewer 4: I don't think `action: DENY, shadowAction: DENY` is a useless state, dynamic metadata and access log hints are only available for the shadow engine AFAIK. See my previous comment.

Reviewer: > dynamic metadata and access log hints are only available

can you expand on what these are and why they might be needed?

Reviewer 4: dynamic metadata is data emitted by filters via `setDynamicMetadata` that can be used down the chain in other filters (see [this](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata)). 

access log hints - This is how an RBAC filter can signal (via dynamic metadata) to other filter that this request should be logged. Side note: this is something that was **not** implemented in Envoy when I worked on TrafficPermissions in Allegro (so I might be wrong in some places).

When we sat down to design TrafficPermissions in Allegro we thought about the end goal of this feature but also about the migration. How would a pre-existing application go from no TrafficPermissions to being "locked-down" via Traffic permissions safely? First of all you need a "grace period" in which the traffic is allowed and metrics / logs are emitted that help to build TrafficPermissions rules. In allegro this was done via logs (that's why I mention access log hints) and stats (we don't have equivalent now I think).

I'm not saying that this feature cannot be released without this but I would be careful about going through any "one way doors" that could significantly complicate implementing things mentioned above.

Reviewer: Do remember we can just add `DENY_AND_SHADOW_DENY`/`ALLOW_AND_SHADOW_ALLOW` at any time.

Reviewer 4: Is this the same with ALLOW/DENY/**LOG**?

Reviewer: Sure, `DENY_AND_SHADOW_LOG` :shrug: 

Btw I think the `access_log_hint` is supported for both actions and shadow actions.

### Envoy configuration

Envoy has [RBAC Network Filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/rbac_filter) 
that allows configuring permissions for the list of principals.

**Principal** is an entity that can be authenticated.<br>
**Permission** is access granted to the principal.  

Each Kuma service has a generated certificate with URIs that identify service name and tags.
For example "backend" service in "us-east" zone of version v1 can have the following list of URIs:
* spiffe://mesh-1/backend
* kuma://kuma.io/zone/us-east
* kuma://version/v1

If we want to grant any permissions to these instances we can create the following Network RBAC:

```yaml
action: ALLOW
policies:
  "MeshTrafficPermission":
    permissions:
      - any: true
    principals:
      - and_ids:
          ids:
            - authenticated:
                principal_name:
                  exact: "spiffe://mesh-1/backend"
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/zone/us-east"
            - authenticated:
                principal_name:
                  exact: "kuma://version/v1"
```

> In order request to be allowed, at least 1 policy should be matched. 

Reviewer 4: 1 policy? do you mean 1 principal?

```suggestion
In order request to be allowed, at least 1 principal should be matched. 
```

                  exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
            - authenticated:
                  principal_name:
                    exact: "kuma://kuma.io/service/web"
      - and_ids:
          ids:
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://kuma.io/zone/us-east"
            - authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/service/web"
      - and_ids:
          ids:
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/service/web"
```
Set this policy as "rules" into NetworkFilter.

> 4. Shadow rules. Take only rules that have "DENY_WITH_SHADOW_ALLOW":

Reviewer 4: Why don't we take `ALLOW_WITH_SHADOW_DENY`?

Author: We don't generate `deny` entries. Deny happens if there is no `allow` entry.

with "ALLOW" action and with a single "MeshTrafficPermission" policy. The policy has permission "any: true" and the list of "principals".
List of principals generated based on the list from step 3. Each principal is matched with OR semantic:

```yaml
action: ALLOW
policies:
  "ShadowMeshTrafficPermission":
    permissions:
      - any: true
    principals:
      - and_ids:
          ids:
            - authenticated:
                principal_name:
                  exact: "kuma://kuma.io/zone/us-east"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://env/dev"
            - not_id:
                authenticated:
                  principal_name:
                    exact: "kuma://kuma.io/service/web"
```
Set this policy as "shadow_rules" into NetworkFilter.

5. If MeshTrafficPermission is targeting proxy then NetworkRBAC is placed as a NetworkFilter on the inbound listener.
If MeshTrafficPermission is targeting MeshHTTPRoute then NetworkRBAC is placed as a NetworkFilter on the route. 

> Even in the case of MeshHTTPRoute, global per-proxy NetworkRBAC is still will be set to provide secure access for other routes. 

Reviewer 4: ```suggestion
Even in the case of MeshHTTPRoute, global per-proxy NetworkRBAC will be set to provide secure access for other routes. 
```

MeshTrafficPermission is going to be the first policy with a new policy matching. 
There are a few things worth mentioning regarding this fact. 

#### Mesh as label

Instead of creating a CRD with a "mesh" field, we can use a label to store mesh:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  labels:
    kuma.io/mesh: mesh-1
```

This allows filtering policies by mesh using kubectl. 
If the label is unset it will be added automatically by the webhook to be "kuma.io/mesh: default".

#### Scope is namespaced

Today all policies are cluster-scoped, but new policies should be namespace scoped.
MeshTrafficPermission can only be created in the namespace where the control-plane runs (`kuma-system` by default), 
and affects DPPs across all namespaces.
In the future we'll add support for other namespaces, so the policy could affect DPPs only in the specified namespace but this is out of scope for this MADR. 

### Cross-mesh

MeshTrafficPermission is used to specify a list of clients from other meshes that have an access to MeshGatewayRoute.
In that case we use targetRef with "mesh" field:

```yaml
type: MeshTrafficPermission

> mesh: expose

Reviewer 5: Do we need a mesh field here? Based on section `Mesh as label` I think we should get rid of it here.

Author: This is a `Universal` format, `Mesh as label` is only for k8s
